Histogram Transport Implementation

Daniel-Packer opened the pull request:
These changes implement the "First Lower Bound" from the paper "Gromov–Wasserstein Distances and the Metric Approach to Object Matching" (Memoli 2011). The algorithm as implemented is faster than the entropic GW computation, and sometimes even outputs a coupling that is more optimal than the one found via entropic GW.

We refer to the method as "Histogram Transport," which is short for "Histogram Transport between Local Distributions of Distances" in the hope of being vaguely descriptive. The actual algorithm is implemented in `src/ott/solvers/quadratic/histogram_transport.py`. Additionally, because of the value added by initializing GW from the Histogram Transport coupling, we have implemented the `FixedCouplingInitializer` to allow GW initialization from any starting point.

Two tests have also been implemented to check that the distance works, and that it has the correct gradients.

Daniel-Packer commented:
Additionally, I have a brief notebook that shows the efficacy of the implementation, which I can't seem to copy/paste here.
I've uploaded an html export of it here:
https://www.asc.ohio-state.edu/packer.61/lower_bound_test.html
And the actual notebook here:
https://www.asc.ohio-state.edu/packer.61/first_lower_bound_tests.ipynb

Let me know if there's a better place for me to put it!

marcocuturi commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,243 @@
+    x_indices = jnp.round(small_indices * (m / min_num_pts)).astype(int)
+    y_indices = jnp.round(small_indices * (n / min_num_pts)).astype(int)
+
+    if self.softness < 0:
```
if it's an epsilon, maybe call it `epsilon_ ...` ?

Daniel-Packer replied:
Sounds good! I'm going to go with `epsilon_1d`, but you have more precise suggestions, I'm happy to hear it.

marcocuturi commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,243 @@
+  the local distributions of distnaces.
+
+  Args:
+  softness: regularization for soft sort. Set to `-1` for normal sorting
```
could be 0.0 for hard sorting?

Daniel-Packer replied:
Sounds good to me!

marcocuturi commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,243 @@
+  Computes the First Lower Bound distance between two distributions.
+  The current implementation requires uniform marginals.
+  If there are an uneven number of points in the two distributions,
+  then we perform a stratified subsample of the distribution of
```
i am not sure a stratified sample is the way to go here. I think something a lot simpler could be to use (soft) quantiles, and pass the quantile levels with a default of, e.g., `jnp.linspace(0,1,num=100)` for instance. 

also, I think we're not taking into account point cloud weights here? (i.e. `a`, `b`). I think it's ok to have a solution that only works for uniform, but this must be more explicit.

Daniel-Packer replied:
- So I actually initially tried with `jax.lax.quantile`, but found that it significantly increased the time (using the default interpolation method).
- We are not taking into account point cloud weights here, which I find pretty sad! I do have an implementation for variable weights based on the implementation in `scipy` [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wasserstein_distance.html), but it drops off in speed relative to entropic GW at around 500 data points, so I didn't put it in this PR.
- But yeah, I will try to make the documentation here more explicit

Daniel-Packer replied:
Update on this: In my most recent implementation with quantiles, it actually went at the same speed. I have no clue what was wrong with my other implementation.

marcocuturi commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,243 @@
+        sorted_dists_yy, y_indices[None, :], -1
+    )
+
+    match self.p:
```
here we should be able to pass a `cost_fn`, even if it's 1-D. Although 1D transport can be solved for general costs, we could stick to `TICost` with the assumption that `h` in that `TIFunction` is convex.

marcocuturi replied:
also, match is only available from python3.10, this is why your code does not work for 3.9.

using a `CostFn` you can get rid of this anyway

Daniel-Packer replied:
That would be a convenient assumption for me to make! I'll implement it with that assumption, and we can extend it later.

Daniel-Packer replied:
Yep! I've factored out the other `match` statements I had and removed all the `p` stuff. I'm going to resolve this conversation.

marcocuturi commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,243 @@
+            )
+        )
+      case _:
+        cost_xy = jnp.power(
```
I think 1D transport should be implemented in `solvers.linear`, largely by taking what you have implemented above, and then vmapped here to obtain `cost_xy`. This might then, for instance, be reused for sliced implementation.

Daniel-Packer replied:
Good idea! I put it in `wasserstein_1d.py`--might make sense to rename it, though.

marcocuturi commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,243 @@
+    [Tuple[np.ndarray, np.ndarray, np.ndarray, "HTState"]], None]
+
+
+class HTOutput(NamedTuple):
```
I don't think it's needed to create a new output class that duplicates a lot of  `GWSinkhornOutput`. We could reuse it.

Daniel-Packer replied:
Good point. I have removed these classes.

marcocuturi reviewed (commented):
Thanks a lot!

a few comments:
- not sure this should be a new quadratic solver class. I think quadratic initializer should be enough? Since there's no randomness, I think the lower bound fits better as an initializer.
- the Wasserstein 1D part needs to be ported to `solvers.linear`. Your first attempt is nice (see comments on quantile), but we definitely want to see any better implementation of 1D that we could push in the future directly impact this lower bound layer without having to reimplement anything.
- Thanks!!

marcocuturi commented:
maybe epsilon_sort ?


On Fri, Oct 20, 2023 at 21:42 Daniel Packer ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/ott/solvers/quadratic/histogram_transport.py
> <https://github.com/ott-jax/ott/pull/444#discussion_r1367427530>:
>
> > +    rng: Random number key.
> +
> +    Returns:
> +    The Histogram Transport output.
> +    """
> +    rng = utils.default_prng_key(rng)
> +
> +    dists_xx = prob.geom_xx.cost_matrix
> +    dists_yy = prob.geom_yy.cost_matrix
> +    m, n = dists_xx.shape[0], dists_yy.shape[0]
> +    min_num_pts = min([m, n])
> +    small_indices = jnp.arange(min_num_pts)
> +    x_indices = jnp.round(small_indices * (m / min_num_pts)).astype(int)
> +    y_indices = jnp.round(small_indices * (n / min_num_pts)).astype(int)
> +
> +    if self.softness < 0:
>
> Sounds good! I'm going to go with epsilon_1d, but you have more precise
> suggestions, I'm happy to hear it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ott-jax/ott/pull/444#discussion_r1367427530>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFMWFTPMPOC4HCEOAYQXNELYALHYBAVCNFSM6AAAAAA6HUUKNCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTMOJQHAYDSOBYGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

michalk8 commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,259 @@
+    )
+
+    if self.cost_fn is None:
+      match self.p:
```
@Daniel-Packer Python doesn't have `match/case`. Also to simplify, I think the user should supply only the `cost_fn` and not `p` (we have implemented `SqPNorm` and `PNormP`, so it's a bit redundant.

Will give you a full review once the all the code can be parsed.

Daniel-Packer replied:
Oh yup. I think Python 3.10 has `match/case`, but that's pretty recent. I'll drop the `p`, good point!

marcocuturi commented:
since this is a solver that can also return a coupling, or potentials, i
think it’s better to call it

solvers.linear.univariate

and to ensure it takes a [n,d] matrix as input. when d=1 this is your
implementation. when d is larger this just runs the univariate matchings
independently, returning sum of couplings and 1d distances. this way this
could work for sliced



On Sun, Oct 22, 2023 at 19:28 Daniel Packer ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/ott/solvers/quadratic/histogram_transport.py
> <https://github.com/ott-jax/ott/pull/444#discussion_r1367941775>:
>
> > +      case 1.0:
> +        cost_xy = jnp.sum(
> +            jnp.abs(sorted_dists_xx[:, None, :] - sorted_dists_yy[None, :, :]),
> +            axis=-1,
> +        )
> +      case 2.0:
> +        cost_xy = jax.lax.sqrt(
> +            jnp.sum(
> +                jnp.square(
> +                    sorted_dists_xx[:, None, :] - sorted_dists_yy[None, :, :]
> +                ),
> +                axis=-1,
> +            )
> +        )
> +      case _:
> +        cost_xy = jnp.power(
>
> Good idea! I put it in wasserstein_1d.py--might make sense to rename it,
> though.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ott-jax/ott/pull/444#discussion_r1367941775>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFMWFTKUP25RWOEVDWYU6Z3YAVJTFAVCNFSM6AAAAAA6HUUKNCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTMOJRGQ3DONBQHA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Daniel-Packer commented:
Regarding making this point:
> not sure this should be a new quadratic solver class. I think quadratic initializer should be enough? Since there's no randomness, I think the lower bound fits better as an initializer.

I think there's some intuition here about what makes a Quadratic Solver that I don't have.

To me, it seems like it should be a solver because there are reasonable applications (see the [notebook](https://www.asc.ohio-state.edu/packer.61/lower_bound_test.html)) where one would use this solver instead of a GW solver, since it is sometimes faster and better at finding an optimal quadratic coupling.

But it seems like this point might be unrelated to the main point that you have in mind about what makes a Quadratic Solver. Let me know!

marcocuturi commented:
the reason this is not a quadratic solver is that ultimately this solves a
linear problem 😅 that is all!




Sent from my iPhone


On Sun, Oct 22, 2023 at 19:36 Daniel Packer ***@***.***>
wrote:

> Regarding making this point:
>
> not sure this should be a new quadratic solver class. I think quadratic
> initializer should be enough? Since there's no randomness, I think the
> lower bound fits better as an initializer.
>
> I think there's some intuition here about what makes a Quadratic Solver
> that I don't have.
>
> To me, it seems like it should be a solver because there are reasonable
> applications (see the notebook
> <https://www.asc.ohio-state.edu/packer.61/lower_bound_test.html>) where
> one would use this solver instead of a GW solver, since it is sometimes
> faster and better at finding an optimal quadratic coupling.
>
> But it seems like this point might be unrelated to the main point that you
> have in mind about what makes a Quadratic Solver. Let me know!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ott-jax/ott/pull/444#issuecomment-1774153722>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFMWFTPG4TUNVT4FQBZCXNLYAVKSNAVCNFSM6AAAAAA6HUUKNCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONZUGE2TGNZSGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Daniel-Packer commented:
> since this is a solver that can also return a coupling, or potentials, i think it’s better to call it solvers.linear.univariate and to ensure it takes a [n,d] matrix as input. when d=1 this is your implementation. when d is larger this just runs the univariate matchings independently, returning sum of couplings and 1d distances. this way this could work for sliced
> […](#)
> On Sun, Oct 22, 2023 at 19:28 Daniel Packer ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In src/ott/solvers/quadratic/histogram_transport.py <[#444 (comment)](https://github.com/ott-jax/ott/pull/444#discussion_r1367941775)>: > + case 1.0: + cost_xy = jnp.sum( + jnp.abs(sorted_dists_xx[:, None, :] - sorted_dists_yy[None, :, :]), + axis=-1, + ) + case 2.0: + cost_xy = jax.lax.sqrt( + jnp.sum( + jnp.square( + sorted_dists_xx[:, None, :] - sorted_dists_yy[None, :, :] + ), + axis=-1, + ) + ) + case _: + cost_xy = jnp.power( Good idea! I put it in wasserstein_1d.py--might make sense to rename it, though. — Reply to this email directly, view it on GitHub <[#444 (comment)](https://github.com/ott-jax/ott/pull/444#discussion_r1367941775)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFMWFTKUP25RWOEVDWYU6Z3YAVJTFAVCNFSM6AAAAAA6HUUKNCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTMOJRGQ3DONBQHA> . You are receiving this because you commented.Message ID: ***@***.***>

I'm a bit confused by this discussion. With its current implementation, the solver only returns a cost. Should I implement it to optionally return couplings and dual variables? Also, can't the sliced implementation be done simply with a `vmap`?

Daniel-Packer commented:
Oh yup! I agree with that point about the linear solver. So I should put it in with the Quadratic Initializers?

marcocuturi commented:
> I'm a bit confused by this discussion. With its current implementation, the solver only returns a cost. Should I implement it to optionally return couplings and dual variables? Also, can't the sliced implementation be done simply with a `vmap`?

yes, to me this is the problem, but maybe you can indeed pass it that way.

ultimately, a `solvers.linear.univariate` solver should return costs / potentials and or transport matrix. Given you are thinking of passing a `epsilon_sort`, this would require being careful to `vmap` (i.e. pass a unified number of iterations, as discussed in https://github.com/google-research/group_testing)

but it's ok, i think this can be left for future work :)

michalk8 commented on tests/solvers/quadratic/ht_test.py:
```diff
@@ -0,0 +1,129 @@
+    """Test basic computations point clouds."""
+    geom_x = pointcloud.PointCloud(self.x)
+    geom_y = pointcloud.PointCloud(self.y)
+    tau_a, tau_b = (1.0, 1.0)
```
No need for parentheses.

michalk8 commented on tests/solvers/quadratic/ht_test.py:
```diff
@@ -0,0 +1,129 @@
+        max_iterations=100,
+        epsilon_sort=-1,
+        cost_fn=PNormP(2.0),
+        method="quantile",
```
I think this function should be parametrized by:
- different `epsilon_sort`s
- different methods
- different cost functions

michalk8 commented on tests/solvers/quadratic/ht_test.py:
```diff
@@ -0,0 +1,129 @@
+
+  @pytest.mark.fast()
+  def test_ht_pointcloud(self):
+    """Test basic computations point clouds."""
```
Would remove these comments, they don't add much information.

michalk8 commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,95 @@
+  `epsilon`: regularization parameter for the resulting Sinkhorn problem
+  `min_iterations`: minimum iterations for computing Sinkhorn distance
+  `max_iterations`: maximum iterations for computing Sinkhorn distance
+  `**kwargs`: keyword arguments for the 1D Wasserstein computation
```
This shouldn't contain `**`, should be just `kwargs`.

michalk8 commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,95 @@
+        max_iterations=max_iterations, min_iterations=min_iterations
+    )
+    wass_solver_1d = wasserstein_1d.WassersteinSolver_1d(**kwargs)
+    self.wass_solver_1d_vmap = jax.jit(
```
It would be also good to test the performance of the doubly-vmapped version vs having it batched as in POT.

michalk8 commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,95 @@
+
+  def __init__(
+      self,
+      epsilon: float = 1.0,
```
`epsilon: Optional[float] = None`

michalk8 commented on tests/solvers/quadratic/ht_test.py:
```diff
@@ -0,0 +1,129 @@
+          epsilon=1.0, max_iterations=100
+      )
+
+      lin_solver = sinkhorn.Sinkhorn(
```
I think it's better to instantiate another `HistogramTransport` rather than modifying the old one.

michalk8 commented on tests/solvers/quadratic/ht_test.py:
```diff
@@ -0,0 +1,129 @@
+
+    for i, implicit in enumerate([True, False]):
+      grad_matrices[i] = reg_ht_grad(x, y, self.a, self.b, implicit)
+      assert not jnp.any(jnp.isnan(grad_matrices[i][0]))
```
Consider comparing the gradients with the ones from finite differences, see, e.g., [here](https://github.com/ott-jax/ott/blob/main/tests/solvers/linear/sinkhorn_diff_test.py#L423-L440).

michalk8 replied:
Also it's better to check using `np.testing.assert_array_equal(jnp.isnan(), False)`,
since it gives you a better error message + more info as to which points have NaN grads.

michalk8 commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,95 @@
+
+# @jax.tree_util.register_pytree_node_class
+class HistogramTransport:
+  """Histogram Transport solver.
```
Please add citation to `docs/references.bib` and link to it here.

Daniel-Packer commented on src/ott/initializers/quadratic/initializers.py:
```diff
@@ -23,15 +23,19 @@
 
   Args:
-    kwargs: Keyword arguments.
+  kwargs: Keyword arguments.
```
Ah, my bad. I think my auto-formatter did this.

Daniel-Packer replied:
This should be fixed everywhere now.

Daniel-Packer commented on src/ott/initializers/quadratic/initializers.py:
```diff
@@ -170,3 +177,65 @@ def _create_geometry(
+    h1, h2 = quad_prob.quad_loss
+    tmp1 = h1.func(geom_xx.cost_matrix)
+    tmp2 = h2.func(geom_yy.cost_matrix)
+    # TODO: Add checks that the coupling is valid
```
Sounds good to me!

Daniel-Packer commented on src/ott/initializers/quadratic/initializers.py:
```diff
@@ -170,3 +177,65 @@ def _create_geometry(
+    tmp1 = h1.func(geom_xx.cost_matrix)
+    tmp2 = h2.func(geom_yy.cost_matrix)
+    # TODO: Add checks that the coupling is valid
+    tmp = tmp1 @ self.init_coupling @ tmp2.T
```
hmm... I think these are equivalent/my implementation in the `FixedCouplingInitializer` is bad (in particular, it assumes uniform marginals). We should just be able to use the one in `QuadraticInitializer`, I think. I'll do a test to see if it matters or not real quick and report back.

Daniel-Packer commented on src/ott/initializers/quadratic/initializers.py:
```diff
@@ -170,3 +177,65 @@ def _create_geometry(
     )
+
+
+class FixedCouplingInitializer(BaseQuadraticInitializer):
```
Since we're merging `FixedCouplingInitializer` with `QuadraticInitializer`, I'm going to resolve this issue.

Daniel-Packer commented on src/ott/solvers/linear/wasserstein_1d.py:
```diff
@@ -0,0 +1,88 @@
+from ott.tools import soft_sort
+
+
+class WassersteinSolver_1d:
```
What's the purpose of the `pytree` structure? I want to make sure that I implement the flattening/unflattening correctly.

Daniel-Packer commented on src/ott/solvers/linear/wasserstein_1d.py:
```diff
@@ -0,0 +1,88 @@
+
+  Args:
+  `epsilon_sort`: regularization parameter for sorting. 0.0 for hard-sorting.
+  `cost_fn`: The cost function for transport. Defaults to Euclidean distance.
```
I think that's actually what I implemented without realizing it! My bad.

Daniel-Packer commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,95 @@
+      **kwargs,
+  ):
+    self.epsilon = epsilon
+    self.linear_ot_solver = sinkhorn.Sinkhorn(
```
Sounds good to me!

Daniel-Packer commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,95 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import (
+    Optional,
```
I think we no longer need `Optional` in this class, but I've internalized the more general point.

Daniel-Packer commented on src/ott/solvers/linear/wasserstein_1d.py:
```diff
@@ -0,0 +1,88 @@
+
+
+class WassersteinSolver_1d:
+  """One-Dimensional Optimal Transport Solver.
```
This got changed in the larger refactor to `Univariate`, let me know what your thoughts are.

Daniel-Packer commented on src/ott/solvers/linear/wasserstein_1d.py:
```diff
@@ -0,0 +1,88 @@
+class WassersteinSolver_1d:
+  """One-Dimensional Optimal Transport Solver.
+
+  WARNING: As implemented, this solver assumes uniform marginals,
```
I think I've done this appropriately. Let me know if not!

marcocuturi commented on src/ott/initializers/quadratic/initializers.py:
```diff
@@ -88,7 +90,7 @@ def tree_unflatten(  # noqa: D102
 
 class QuadraticInitializer(BaseQuadraticInitializer):
-  r"""Initialize a linear problem locally around :math:`ab^T` initializer.
+  r"""Initialize a linear problem locally around an initial coupling.
```
double "initial" in same sentence a bit cryptic

marcocuturi commented on src/ott/initializers/quadratic/initializers.py:
```diff
@@ -112,20 +114,28 @@ class QuadraticInitializer(BaseQuadraticInitializer):
+
+  Args:
+    init_coupling: The coupling to use for initialization. Set to `None` for
+    the uniform coupling.
```
2 spaces missing before "the uniform"

marcocuturi commented on src/ott/solvers/linear/univariate.py:
```diff
@@ -0,0 +1,103 @@
+    """Computes the 1D Wasserstein Distance between `x` and `y`."""
+    if self.method == "subsample":
+      return self.cost_fn.pairwise(
+          self._sort(x)[jnp.linspace(0, x.shape[0],
```
if one passes weights, can we imagine doing stratified sampling using [digitize](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.digitize.html)

marcocuturi replied:
(I had in mind this specific conversation: https://github.com/google/jax/issues/2595)

marcocuturi commented on src/ott/solvers/linear/univariate.py:
```diff
@@ -0,0 +1,103 @@
+    if self.require_sort:
+      return x
+    if self.epsilon_sort > 0 or self.epsilon_sort is None:
+      return soft_sort.sort(x, epsilon=self.epsilon_sort)
```
for this to be vmappable, we will need to pass `min_iterations=max_iterations`

marcocuturi reviewed (commented):
Thanks Daniel again! getting better! 

The univariate solver is a tricky beast, so thanks for taking a stab.

Either we want to have 1. an approx (through e.g. quantiles, subsampling etc..) of the OT cost, and nothing else or 2. we do want to have an OT coupling (sparse if no epsilon). 

In case 1, This should also work for various `a`/`b`. I think stratified sampling should work. It seems `quantiles` doesn't implement this, so maybe raise an error if `a,b` are not None in that case (we'll use that to assume the weights are uniform).

In case 2, we'll think about this more deeply after your PR.

For now I think you can go as you did, i.e. only output OT cost. But I think it should be ok to add in the sampling approach something that looks like a `choice` sampler based on weights, and then

michalk8 commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,96 @@
+
+
+@jax.tree_util.register_pytree_node_class
+class HistogramTransport:
```
The latter sounds better, would maybe drop the `Memoli` + add citation in the docs.

michalk8 commented on src/ott/solvers/linear/univariate.py:
```diff
@@ -0,0 +1,103 @@
+    raise KeyError(f"Method {self.method} not implemented!")
+
+  def _sort(self, x: jnp.ndarray) -> jnp.ndarray:
+    if self.require_sort:
```
Should be `if not self.require_sort: return x`

michalk8 commented on src/ott/solvers/linear/univariate.py:
```diff
@@ -0,0 +1,103 @@
+                                     num=self.n_subsamples).astype(int)],
+          self._sort(y)[jnp.linspace(0, y.shape[0],
+                                     num=self.n_subsamples).astype(int)],
+      )
```
I'd either rename `_sort` to `_maybe_sort` (less preferred) or do the following:
```python
sorted_x, sorted_y = (self._sort(x), self._sort(y)) if self.requires_sort else (x, y)
if self.method == "equal":
   return self.cost_fn.pairwise(sorted_x, sorted_y)
...
```

michalk8 commented on src/ott/solvers/linear/univariate.py:
```diff
@@ -0,0 +1,103 @@
+      epsilon_sort: Optional[float] = 0.0,
+      cost_fn: Optional[costs.TICost] = None,
+      method: Literal["subsample", "quantile", "equal"] = "subsample",
+      n_subsamples: int = 100,
```
Am pondering whether this should have a default or not (None), wdyt @marcocuturi ?

michalk8 commented on src/ott/solvers/linear/univariate.py:
```diff
@@ -0,0 +1,103 @@
+  def __init__(
+      self,
+      epsilon_sort: Optional[float] = 0.0,
+      cost_fn: Optional[costs.TICost] = None,
```
Based on discussion with Marco, please switch back to `costs.CostFn`.

michalk8 commented on src/ott/solvers/linear/univariate.py:
```diff
@@ -0,0 +1,103 @@
+
+  def tree_flatten(self):  # noqa: D102
+    aux = vars(self).copy()
+    return [], aux
```
I think `epsilon_sort` and the `cost_fn` can be children.

michalk8 commented on src/ott/solvers/quadratic/histogram_transport.py:
```diff
@@ -0,0 +1,96 @@
+class HistogramTransport:
+  """Histogram Transport solver.
+
+  WARNING: As implemented, this solver assumes uniform marginals,
```
Should be:
```rst
.. warning::
  As implemented, this solver ...
```

Daniel-Packer commented:
Okay--I think I have addressed most comments, although some things may have slipped through the cracks.

An important thing I have _not_ figured out:

- I can't get the implementation to pass any differentiability tests. I suspect that this is for two reasons:
  1. The implementation with hard-sorting is not differentiable
  2. I don't really know how to make sure that jax's implicit differentiation tracks everything correctly.

Otherwise, what other issues do y'all want resolved?

Daniel-Packer commented:
Okay, I have made all the requested changes aside from the obj:`None` one, which I don't think I understood. Otherwise, we still need to decide if `solvers.quadratic` is where the solver should live. And also what name to give it.

michalk8 commented:
@Daniel-Packer I've made some modifications how sorting function is passed to the `UnivariateSolver`, I think it cleans much of the arguments!.
Furthermore, I've renamed the `HistogramTransport` -> `LowerBoundSolver` per our discussions and added a differentiability tests.

From my side it's ready to be merged once all the tests pass, thanks a lot for the PR!
If you want, would be great to include the notebook that you've linked in a separate PR (after this PR is merged).
Alternatively, we can add a new section to the [GW notebook](https://ott-jax.readthedocs.io/en/latest/tutorials/gromov_wasserstein.html); slightly prefer the latter, but pinging @marcocuturi .

Daniel-Packer commented:
@michalk8 Thank you so much for all the edits. I'll add a PR for editing the GW notebook once this one is merged. Also, I think there might be another PR for a weighted univariate solver too upcoming.
